File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_6.java
package org.twyszomirski.aoc.solutions;

import org.twyszomirski.aoc.utils.Tuple;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

import static java.io.IO.println;

public class Day_6 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_6.txt"));
        println("======== Day 6 ===========");
        part_1(lines);
        part_2(lines);
    }

    private static void part_1(List<String> lines) {
        var data = readData(lines);
        var start = data.stream().flatMap(Collection::stream)
                .filter(MatrixPoint::isStart).findFirst().get();

        var result = start.move(data, Direction.UP);
        Set<MatrixPoint> points = new HashSet<>();
        points.add(start);
        while (result.left() != null) {
            points.add(result.left());
            result = result.left().move(data, result.right());
        }

        println(points.size());
    }

    private static void part_2(List<String> lines) {
        var data = readData(lines);

        long count = 0;
        for (int row = 0; row < lines.size(); row++) {
            for (int col = 0; col < data.get(row).size(); col++) {
                var copy = copyWithObstruction(data, row, col);
                if (copy != null && hasCycle(copy)) {
                    count++;
                }
            }
        }
        println(count);
    }


    private static List<List<MatrixPoint>> readData(List<String> lines) {
        List<List<MatrixPoint>> matrix = new ArrayList<>();
        for (int row = 0; row < lines.size(); row++) {
            var r = Arrays.stream(lines.get(row).split("")).filter(it -> !it.isBlank()).toList();
            var c = new ArrayList<MatrixPoint>();
            for (int column = 0; column < r.size(); column++) {
                c.add(new MatrixPoint(row, column, r.get(column)));
            }
            matrix.add(c);
        }
        return matrix;
    }

    record MatrixPoint(Integer row, Integer column, String symbol) {

        public boolean isObstacle() {
            return "#".equals(symbol);
        }

        public boolean isStart() {
            return "^".equals(symbol);
        }

        public Tuple<MatrixPoint, Direction> move(List<List<MatrixPoint>> matrix, Direction direction) {
            var row = this.row + direction.vertical;
            var column = this.column + direction.horizontal;
            MatrixPoint next;
            if (row >= 0 && row < matrix.size() && column >= 0 && column < matrix.getFirst().size()) {
                next = matrix.get(row).get(column);
            } else {
                return new Tuple<>(null, direction);
            }
            if (next.isObstacle()) {
                return this.move(matrix, direction.rotate());
            }

            return new Tuple<>(next, direction);
        }
    }

    private static List<List<MatrixPoint>> copyWithObstruction(List<List<MatrixPoint>> initial, Integer row, Integer column) {
        MatrixPoint originalPoint = initial.get(row).get(column);
        if (originalPoint.isObstacle() || originalPoint.isStart()) {
            return null;
        }
        return initial.stream()
                .map(r -> r.stream()
                        .map(p -> (p.row.equals(row) && p.column.equals(column)) ? new MatrixPoint(p.row, p.column, "#") : p)
                        .toList())
                .toList();
    }

    private static boolean hasCycle(List<List<MatrixPoint>> matrix) {
        var start = matrix.stream().flatMap(Collection::stream)
                .filter(MatrixPoint::isStart).findFirst().get();
        var direction = Direction.UP;
        var fastResult = new Tuple<MatrixPoint, Direction>(start, direction);
        var slowResult = new Tuple<MatrixPoint, Direction>(start, direction);

        var hasCycle = false;

        while (fastResult.left() != null) {
            fastResult = fastResult.left().move(matrix, fastResult.right());
            if (fastResult.left() == null) {
                break;
            }
            fastResult = fastResult.left().move(matrix, fastResult.right());
            slowResult = slowResult.left().move(matrix, slowResult.right());

            if (fastResult.left() == slowResult.left() && fastResult.right() == slowResult.right()) {
                hasCycle = true;
                break;
            }
        }
        return hasCycle;
    }

    enum Direction {
        UP(-1, 0),
        RIGHT(0, 1),
        DOWN(1, 0),
        LEFT(0, -1);

        private Integer vertical;
        private Integer horizontal;


        Direction(Integer vertical, Integer horizontal) {
            this.vertical = vertical;
            this.horizontal = horizontal;
        }

        public Direction rotate() {
            return switch (this) {
                case UP -> RIGHT;
                case RIGHT -> DOWN;
                case DOWN -> LEFT;
                case LEFT -> UP;
            };
        }

    }
}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_6.kt
package org.twyszomirski.aoc.solutions

import java.io.File


class Day_6 {

    fun solve() {
        val lines = File("src/main/resources/input_day_6.txt").readLines()
        println("======== Day 6 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val matrix = readMatrix(input)
        val direction = Direction(0, -1)
        val start = matrix.flatten().first { it.visited }

        var result = start to direction
        while (result.first.isInsideMatrix(matrix)) {
            result = result.first.moveOnMatrix(matrix, result.second)
        }
        val sum = matrix.flatten().count { it.visited }
        println(sum)
    }

    fun part2(input: List<String>) {
        val initial = readMatrix(input)
        val direction = Direction(0, -1)
        val start = initial.flatten().first { it.visited }

        val count = initial.flatten().map {
            var cycle = false
            if (!it.obstacle && it != start) {
                val withAdditionalObstacle = readMatrix(input, it)

                var resultSlow = start to direction
                var resultFast = start to direction

                while (resultFast.first.isInsideMatrix(initial)) {
                    resultFast = resultFast.first.moveOnMatrix(withAdditionalObstacle, resultFast.second)
                    resultFast = resultFast.first.moveOnMatrix(withAdditionalObstacle, resultFast.second)

                    resultSlow = resultSlow.first.moveOnMatrix(withAdditionalObstacle, resultSlow.second)

                    if (resultSlow.first == resultFast.first && resultSlow.second == resultFast.second) {
                        cycle = true
                        break
                    }
                }
            }
            cycle
        }.count { it }
        println(count)
    }

    fun readMatrix(input: List<String>, additionalObstacle: Point? = null): List<List<Point>> {
        return input.mapIndexed { row, data ->
            data.split("").filter { it.isNotBlank() }
                .mapIndexed { col, data ->
                    val type = if (additionalObstacle?.x == col && additionalObstacle?.y == row) "#" else data
                    Point.ofData(col, row, type)
                }
        }
    }

}

data class Point(val x: Int, val y: Int, val obstacle: Boolean = false, var visited: Boolean = false) {

    fun isInsideMatrix(matrix: Matrix) = x >= 0 && y >= 0 && x < matrix.first().size && y < matrix.size

    fun moveOnMatrix(matrix: Matrix, direction: Direction): Pair<Point, Direction> {
        val newPoint = Point(x + direction.horizontal, y + direction.vertical)
        if (newPoint.isInsideMatrix(matrix)) {
            val onMatrix = matrix[newPoint.y][newPoint.x]
            if (onMatrix.obstacle) {
                return moveOnMatrix(matrix, direction.rotated())
            } else {
                onMatrix.visited = true
                return Pair(onMatrix, direction)
            }
        }
        return Pair(newPoint, direction)
    }

    companion object {
        fun ofData(x: Int, y: Int, data: String): Point {
            return when (data) {
                "." -> Point(x, y, false)
                "#" -> Point(x, y, true)
                "^" -> Point(x, y, false, true)
                else -> throw RuntimeException("Unknown point.")
            }
        }
    }
}

data class Direction(val horizontal: Int, val vertical: Int) {
    fun rotated(): Direction {
        return when (this) {
            Direction(0, 1) -> Direction(-1, 0)
            Direction(-1, 0) -> Direction(0, -1)
            Direction(0, -1) -> Direction(1, 0)
            Direction(1, 0) -> Direction(0, 1)
            else -> throw RuntimeException("Not in this universe")
        }
    }
}

typealias Matrix = List<List<Point>>